### PR TITLE
Windows time accuracy

### PR DESCRIPTION
**Pull request by danielhrisca**
Improve time stamp accuracy of the CAN messages for Kvaser and Vector buses on the Windows platform. (see #934)

On Windows the value returned by `time.time` is refreshed at best every 0.5ms, and on some PCs the default refresh rate is every 10ms.

The idea is to get the value of `time.perf_counter` at the moment when `time.time` is refreshed and use this as a reference point.

### CODE REVIEW

File: can/util.py [hunk 1]
```diff
@@ -326,6 +327,26 @@ def rename_kwargs(func_name, kwargs, aliases):
                 warnings.warn("{} is deprecated".format(alias), DeprecationWarning)
 
 
+def time_perfcounter_correlation():
+    """ on Windows platform this will help with the accuracy of using the time.time fucntion.
```
Comment (hardbyte) [inline can/util.py:331]:
```suggestion
    """Get the `perf_counter` value nearest to when time.time() is updated on Windows.
    
    On non Windows platforms the current time and perf_counter is directly returned.
    
    Note this value is based on when `time.time()` is observed to update from Python, it is not directly returned by the operating system.
```

File: can/util.py [hunk 2]
```diff
@@ -326,6 +327,26 @@ def rename_kwargs(func_name, kwargs, aliases):
...
+
+    Returns
+    -------
+    t, performance_counter : (time.time, float)
+        time.time value and perf_counter value when the time.time is updated
+
+    """
```
Comment (hardbyte) [inline can/util.py:338]:
```suggestion
    :return: (t, performance_counter) time.time value and perf_counter value when the time.time is updated
    """
```

File: can/util.py [hunk 3]
```diff
@@ -326,6 +327,26 @@ def rename_kwargs(func_name, kwargs, aliases):
...
+    if platform.system() == "Windows":
+        t0 = time()
+        while 1:
```
Comment (hardbyte) [inline can/util.py:341]:
```suggestion
        while True:
```

**Review by hardbyte — APPROVED** [review]
Thanks for the PR. I'd like a basic test of the added utility function - just check it returns results of the expected types and ensure it always returns within say 500ms. 

Might be worth looking at the docs for those interfaces and adding a note too.

Is there any harm in removing the platform check and including this correction for all operating systems? It can only help other platforms right?

**Comment by christiansandberg** [comment]
Or query the resolution using `get_clock_info`. If it's too low do the compensation.

**Comment by zariiii9003** [comment]
Don't you have the same `time()`  inaccuracies in the correlation step `t1, performance_counter = time(), perf_counter()`?

**Review by hardbyte — APPROVED** [review]
Good stuff.

🕐

**Review by christiansandberg — APPROVED** [review]

File: can/interfaces/vector/canlib.py [hunk 4]
```diff
@@ -295,11 +300,22 @@ def __init__(
         # Calculate time offset for absolute timestamps
         offset = xlclass.XLuint64()
         try:
-            try:
-                xldriver.xlGetSyncTime(self.port_handle, offset)
-            except VectorError:
-                xldriver.xlGetChannelTime(self.port_handle, self.mask, offset)
-            self._time_offset = time.time() - offset.value * 1e-9
+            if time.get_clock_info("time").resolution > 1e-5:
```
Comment (rliebscher) [inline can/interfaces/vector/canlib.py:303]: There is a lot of code duplication. Why not wrapping it in an object like this in can_util:
```
class PreciseTime:
    def __init__(self):
            # use this if the resolution is higher than 10us
            if get_clock_info("time").resolution > 1e-5:
                t0 = time()
                while True:
                    t1, performance_counter = time(), perf_counter()
                    if t1 != t0:
                        break
                self.time = t1
                self.perfcounter = performance_counter
            else:
                self.time = time()
                self.perfcounter = None # not needed

    def now(self):
        if self.perf_counter is None:
            return time()
        return self.time + (perf_counter() - self.perfcounter)
```
Then you need no extra if to you use it, you could just write:
```
            precisetime = PreciseTime()
            try:
                xldriver.xlGetSyncTime(self.port_handle, offset)
            except VectorError:
                xldriver.xlGetChannelTime(self.port_handle, self.mask, offset)
            self._time_offset = precisetime.now() - offset.value * 1e-9
```

Comment (rliebscher) [reply can/interfaces/vector/canlib.py:303]: BTW: the constructor could have also a param required_resolution instead of hard coding 1e-5

**Comment by danielhrisca** [comment]
What should I do next to get the ball rolling again for this PR?

**Comment by hardbyte** [comment]
I'll merge it now, mergify usually takes over once there are 2 approvals and the CI has passed - not sure why it didn't this time.

**Pull request merged**